sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+the front-end each time we add a new property, so the front-end is not required
+to understand each property it consumes.  Instead, it follows a mechanical
+process for converting properties listed in the C++ annotation into LLVM IR,
+and this mechanical process need not be updated when we add new properties.
+
+Once the information about properties is represented in IR, it is available to
+compiler passes.  For example, the `sycl-post-link` tool might use a property
+in order to perform one of its transformations.  Some properties are consumed
+by the DPC++ compiler, but others are transformed into SPIR-V for use by the
+JIT compiler.  This design document also describes how this SPIR-V
+transformation is done.
+
+
+## Properties on a global variable type
+
+One use for compile-time properties is with types that are used exclusively
+for declaring global variables.  One such example is the
+[SYCL\_EXT\_ONEAPI\_DEVICE\_GLOBAL][2] extension:
+
+[2]: <extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc>
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The following code illustrates a `device_global` variable that is declared with
+two compile-time properties:
+
+```
+using sycl::ext::oneapi;
+
+device_global<int,
+  property_list_t<
+    device_image_scope::value_t,
+    host_access::value_t<host_access::access::read>>>
+  dm1;
+```
+
+The header file represents these properties with an internal C++ attribute
+named `[[__sycl_detail__::add_ir_global_variable_attributes()]]` whose value
+is a list that is created through a template parameter pack expansion:
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  device_global<T, property_list<Props...>> {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The initial entries in the C++ attribute's parameter list are the names of the
+properties, and these are followed by the values of the properties.  To
+illustrate using the same example as before, the result of the parameter pack
+expansion would look like this:
+
+```
+namespace sycl::ext::oneapi {
+
+template </* ... */> class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    "sycl-device-image-scope",  // Name of first property
+    "sycl-host-access",         // Name of second property
+    "",                         // First property has no parameter
+    "read"                      // Value of second property
+    )]]
+#endif
```

Comment:
How does one know when the property list ends and the value list begins?

Reply:
I was thinking that the CFE would count the total number of parameters and divide by two.  This position marks the first parameter that is a "value".

Reply:
Is it guaranteed somewhere that each property has one and only one value?   That would make this possible.

Reply:
Yes, this was my intent.  I see that this is not stated very clearly, though, so I can update the spec to make this more clear.

Reply:
Clarified in 44763693264bc91488fc5c8772b61f0cc5d9210b.